dll/win32/shlwapi/autocomp.c

```diff
@@ -0,0 +1,671 @@
+    HKEY hKey;
+    LONG result;
+    DWORD i, cbValue;
+    WCHAR szName[2], szMRUList[64], szValue[MAX_PATH + 32];
+    INT cch;
+
+    result = RegOpenKeyExW(HKEY_CURRENT_USER, pszRunMRU, 0, KEY_READ, &hKey);
+    if (result != ERROR_SUCCESS)
+    {
+        TRACE("Opening RunMRU failed: 0x%lX\n", result);
+        return;
+    }
+
+    szMRUList[0] = 0;
+    cbValue = sizeof(szMRUList);
+    result = RegQueryValueExW(hKey, L"MRUList", NULL, NULL, (LPBYTE)szMRUList, &cbValue);
+    if (result == ERROR_SUCCESS)
+    {
+        for (i = 0; i <= L'z' - L'a'; ++i)
+        {
+            if (szMRUList[i] == 0)
+                break;
+
+            szName[0] = szMRUList[i];
+            szName[1] = 0;
+            szValue[0] = 0;
+            cbValue = sizeof(szValue);
+            result = RegQueryValueExW(hKey, szName, NULL, NULL, (LPBYTE)szValue, &cbValue);
+            if (result != ERROR_SUCCESS)
+                continue;
+
+            cch = wcslen(szValue);
+            if (cch >= 2 && wcscmp(&szValue[cch - 2], L"\\1") == 0)
+            {
+                szValue[cch - 2] = 0;
+            }
+
+            if (UrlIsW(szValue, URLIS_URL))
+            {
+                AutoComplete_EnumString_AddString(pES, szValue);
+            }
+        }
+    }
+
+    RegCloseKey(hKey);
+}
+
+static HRESULT STDMETHODCALLTYPE
+AutoComplete_EnumString_Reset(IEnumString* This)
+{
+    DWORD attrs;
+    WCHAR szText[MAX_PATH];
+    AutoComplete_EnumString *this_ = (AutoComplete_EnumString *)This;
+    DWORD dwSHACF = this_->m_dwSHACF;
+
+    GetWindowTextW(this_->m_hwndEdit, szText, ARRAYSIZE(szText));
+    attrs = GetFileAttributesW(szText);
+
+    AutoComplete_EnumString_ResetContent(this_);
+
+    if (dwSHACF & (SHACF_FILESYS_ONLY | SHACF_FILESYSTEM | SHACF_FILESYS_DIRS))
+    {
+        BOOL bDirOnly = !!(dwSHACF & SHACF_FILESYS_DIRS);
+        if (attrs != INVALID_FILE_ATTRIBUTES)
+        {
+            if (attrs & FILE_ATTRIBUTE_DIRECTORY)
+            {
+                AutoComplete_DoDir(this_, szText, bDirOnly);
+            }
+        }
+        else if (szText[0])
+        {
+            PathRemoveFileSpecW(szText);
+            AutoComplete_DoDir(this_, szText, bDirOnly);
+        }
+        else
+        {
+            AutoComplete_DoDrives(this_, bDirOnly);
+        }
+    }
+
+    if (!(dwSHACF & (SHACF_FILESYS_ONLY)))
+    {
+        if (dwSHACF & SHACF_URLHISTORY)
+        {
+            AutoComplete_DoURLHistory(this_);
+        }
+        if (dwSHACF & SHACF_URLMRU)
+        {
+            AutoComplete_DoURLMRU(this_);
+        }
+    }
+
+    this_->m_istr = 0;
+    return S_OK;
+}
+
+static AutoComplete_EnumString *
+AutoComplete_EnumString_Construct(SIZE_T capacity)
+{
+    AutoComplete_EnumStringVtbl *lpVtbl;
+    AutoComplete_EnumString *ret = CoTaskMemAlloc(sizeof(AutoComplete_EnumString));
+    if (!ret)
+        return ret;
+
+    lpVtbl = CoTaskMemAlloc(sizeof(AutoComplete_EnumStringVtbl));
```

Comment:
The vtable can just be static, compile-time initialized?

Reply:
OK.